# Projeto 1 - Ciência dos Dados

Caio Monteiro Marquez Barbosa
Felipe Bakowski Nantes de Souza
Lucas Santos Teixeira

Atenção: Serão permitidos grupos de três pessoas, mas com uma rubrica mais exigente. Grupos deste tamanho precisarão fazer um questionário de avaliação de trabalho em equipe

___
Carregando algumas bibliotecas:

In [43]:
%matplotlib inline
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import os

In [44]:
print('Esperamos trabalhar no diretório')
print(os.getcwd())

Esperamos trabalhar no diretório
c:\Users\Felipe\Desktop\CdaDos_P1


Carregando a base de dados com os tweets classificados manualmente:

In [45]:
train = pd.read_excel('dados_treino.xlsx')
train.head(5)

Mensagem  Unnamed: 1  Unnamed: 2  \
0  Um dos livros faltam mais de 40 páginas. Ex.: ...         NaN         NaN   
1  O livro é incompleto. Acaba em 11/04/44, sendo...         NaN         NaN   
2  Ainda não terminei, estou nos 40%, devo ler em...         NaN         NaN   
3  Pedi o livro baseado na promoção q era de graç...         NaN         NaN   
4  Prezados, Acredito que houve um erro, pois des...         NaN         NaN   

   Unnamed: 3  Unnamed: 4  Unnamed: 5  Unnamed: 6  Unnamed: 7  Unnamed: 8  \
0         NaN         NaN         NaN         NaN         NaN         NaN   
1         NaN         NaN         NaN         NaN         NaN         NaN   
2         NaN         NaN         NaN         NaN         NaN         NaN   
3         NaN         NaN         NaN         NaN         NaN         NaN   
4         NaN         NaN         NaN         NaN         NaN         NaN   

   Unnamed: 9  ...  Unnamed: 87  Unnamed: 88  Unnamed: 89  Unnamed: 90  \
0         NaN  ...          NaN          NaN          NaN          NaN   
1         NaN  ...          NaN          NaN          NaN          NaN   
2         NaN  ...          NaN          NaN          NaN          NaN   
3         NaN  ...          NaN          NaN          NaN          NaN   
4         NaN  ...          NaN          NaN          NaN          NaN   

   Unnamed: 91  Unnamed: 92  Unnamed: 93  Unnamed: 94  Unnamed: 95    Target  
0          NaN          NaN          NaN          NaN          NaN   entrega  
1          NaN          NaN          NaN          NaN          NaN   entrega  
2          NaN          NaN          NaN          NaN          NaN  conteúdo  
3          NaN          NaN          NaN          NaN          NaN   entrega  
4          NaN          NaN          NaN          NaN          NaN   entrega  

[5 rows x 97 columns]

In [46]:
test = pd.read_excel('dados_teste.xlsx')
test.head(5)


Mensagem  Unnamed: 1  Unnamed: 2  \
0  a narrativa é muito lenta e você fica ansioso ...         NaN         NaN   
1  Sim.adorei o livro Vou continuar a comprar o r...         NaN         NaN   
2  Desculpem, mas achei deprimente e não me ensin...         NaN         NaN   
3  Preciso saber se essa compra é do livro físico...         NaN         NaN   
4  Embora tenha gostado do livro "Não Faz Sentido...         NaN         NaN   

   Unnamed: 3  Unnamed: 4  Unnamed: 5  Unnamed: 6  Unnamed: 7  Unnamed: 8  \
0         NaN         NaN         NaN         NaN         NaN         NaN   
1         NaN         NaN         NaN         NaN         NaN         NaN   
2         NaN         NaN         NaN         NaN         NaN         NaN   
3         NaN         NaN         NaN         NaN         NaN         NaN   
4         NaN         NaN         NaN         NaN         NaN         NaN   

   Unnamed: 9  ...  Unnamed: 33  Unnamed: 34  Unnamed: 35 Unnamed: 36  \
0         NaN  ...          NaN          NaN          NaN         NaN   
1         NaN  ...          NaN          NaN          NaN         NaN   
2         NaN  ...          NaN          NaN          NaN         NaN   
3         NaN  ...          NaN          NaN          NaN         NaN   
4         NaN  ...          NaN          NaN          NaN         NaN   

   Unnamed: 37  Unnamed: 38  Unnamed: 39  Unnamed: 40  Unnamed: 41    Target  
0          NaN          NaN          NaN          NaN          NaN  conteúdo  
1          NaN          NaN          NaN          NaN          NaN  conteúdo  
2          NaN          NaN          NaN          NaN          NaN  conteúdo  
3          NaN          NaN          NaN          NaN          NaN   entrega  
4          NaN          NaN          NaN          NaN          NaN  conteúdo  

[5 rows x 43 columns]

___
## Classificador automático


Faça aqui uma descrição do seu assunto e o que considerou como relevante ou não relevante na classificação dos tweets (Target).

ESCREVA AQUI...

___
### Montando um Classificador Naive-Bayes

Considerando apenas as mensagens da planilha Treinamento, ensine  seu classificador.

In [47]:
df_train = train['Mensagem'].to_frame()
df_train['Target'] = train['Target']

df_test = test['Mensagem'].to_frame()
df_test['Target'] = test['Target']




In [48]:
caracteres = [':',',','.','(',')', '?', '!', '"']
i = 0
for msg in df_train.Mensagem:
    for caractere in caracteres:
        if caractere in msg:
            msg = msg.replace(caractere, '')
            msg = msg.lower()
    df_train.loc[i, 'Mensagem'] = msg 
    i += 1
    
    
j = 0
for msg in df_test.Mensagem:
    for caractere in caracteres:
        if caractere in msg:
            msg = msg.replace(caractere, '')
            msg = msg.lower()
    df_test.loc[j, 'Mensagem'] = msg 
    j += 1

In [49]:
df_train.head(5)



Mensagem    Target
0  um dos livros faltam mais de 40 páginas ex pul...   entrega
1  o livro é incompleto acaba em 11/04/44 sendo q...   entrega
2  ainda não terminei estou nos 40% devo ler em m...  conteúdo
3  pedi o livro baseado na promoção q era de graç...   entrega
4  prezados acredito que houve um erro pois desco...   entrega

In [50]:
df_test.head(5)

Mensagem    Target
0  a narrativa é muito lenta e você fica ansioso ...  conteúdo
1  simadorei o livro vou continuar a comprar o re...  conteúdo
2  desculpem mas achei deprimente e não me ensino...  conteúdo
3  preciso saber se essa compra é do livro físico...   entrega
4  embora tenha gostado do livro não faz sentido ...  conteúdo

___
### Verificando a performance do Classificador

Agora você deve testar o seu classificador com a base de Testes.

In [51]:
lista_mensagem_teste = pd.Series(df_train.Mensagem) #pd series transforma em lista

In [52]:
#criando o conjunto universo de palavras
lista_mensagem_teste = pd.Series(df_train.Mensagem)

lista_dos_splits = []
for msg in lista_mensagem_teste:
    for palavra in msg.split():
        lista_dos_splits.append(palavra)


universo = list(set(lista_dos_splits))


print(universo)

print(len(universo))

['gramatura', 'http//wwwamazoncombr/game-thrones-song-fire-ebook/dp/b000qcs8tw/ref=pd_sim_sbs_kinc_6', 'adptou', 'biblioteca', 'cer-imônias', 'ficar', 'correio', 'tio', 'banheiro', 'definição', 'esse', 'sugestões', 'chanceler', 'devolvi', '99', 'noms', 'unidades', 'formado', 'mínimo', 'históriamas', 'prendesse', 'sinceramente', 'record', 'elogios', 'vc', 'demais', 'itatiaia', 'ideológico', 'trocas', 'pestanejar', 'irreal', 'santos', 'twist', 'lyon', 'ele', 'demonstrou', 'personalidades', 'blog', 'originais', 'minorias', 'disse', 'castellaniposso', 'destópica', 'laranja', 'aplausos', 'confortável', "--'", 'resumido', 'formatações', 'wright', 'chiche', 'locais', 'características', 'prolixo', 'allen', 'prestado', 'afirmar', 'enaltecer', 'golpe', 'arranhada/rasgada', 'atrasado', 'positivos', 'compilador', 'odiei', 'compraram', 'rastrearisso', 'for', 'bonito', 'um”influenciador”', 'linguiça', 'menina', 'la', 'absurdas', 'embrulho', 'embalagem', 'previsto', 'ameaças', 'compras', 'justificara

In [53]:
#criando o conjunto de palavras de entrega
livro_conteudo = ''
df_train_conteudo = df_train.loc[(df_train.Target == 'conteúdo')]
for msg in df_train_conteudo.Mensagem:
    livro_conteudo += msg

In [54]:
livro_conteudo

"ainda não terminei estou nos 40% devo ler em mais uns dois dias mudando de opinião retifico o comentário salvo engano auto-ajuda da pior qualidade cheia de auto-elogios auto-referências a produtos e sites do autor escrito principalmente em primeira pessoa eu isso eu aquilo como eu sou bom cheio de contradições o autor começa contando como ele era um vendedor bem sucedido e no último ano de carreira resolveu se tornar o melhor da empresa já em referência futura ele afirma que era um fracassado até esse último ano em que foi o melhor vendedor da empresa curiosamente também não assume responsabilidade afirma que ficou devendo meio milhão de dólares pela crise de 2008 e não porque gastava demais ou era ruim no que fazia sempre há uma justificativa de quanto ele é bom e sabe das coisas e de como você é ruim se não quiser fazer o que ele propõe chega ao absurdo de dizer que se você afirmar antes de dormir que acordará disposto não faz diferença se dormir 3h da mesma forma que se dormir 9h e

In [55]:
livro_entrega = ''
df_train_entrega = df_train.loc[(df_train.Target == 'entrega')]
for msg in df_train_entrega.Mensagem:
    livro_entrega += msg

In [56]:
livro_entrega

"um dos livros faltam mais de 40 páginas ex pula da pág 140 pra pág 193 como foi um presente e a pessoa só viu quando estava lendo essa parte perdemos o prazo de devoluçãoo livro é incompleto acaba em 11/04/44 sendo que o diário dela foi escrito até 01/08/44 alguém pode explicar pq não informam que é incompletopedi o livro baseado na promoção q era de graça não autorizo está compra ajustem o site de vcs para evitar tais desconfortos dória tem razão e vcs não cumpriram com a palavra não aceito pagar r$2100 por este livro pois estava informando q era gratuitoprezados acredito que houve um erro pois descontaram o valor deste livro no meu cartão de crédito quando escolhi este livro foi no dia 30 de março no momento de uma campanha da empresa de doação de ebooks e este livro era uns dos livros disponíveis gratuitamente espero uma solução rápida pois não gostaria de ser lesadoVeio amassado nas pontasoh amazon por favor atualiza essa edição está sem condição de ler faltam frases dos personage

In [57]:
livro_geral = livro_entrega + livro_conteudo

In [58]:
livro_geral

"um dos livros faltam mais de 40 páginas ex pula da pág 140 pra pág 193 como foi um presente e a pessoa só viu quando estava lendo essa parte perdemos o prazo de devoluçãoo livro é incompleto acaba em 11/04/44 sendo que o diário dela foi escrito até 01/08/44 alguém pode explicar pq não informam que é incompletopedi o livro baseado na promoção q era de graça não autorizo está compra ajustem o site de vcs para evitar tais desconfortos dória tem razão e vcs não cumpriram com a palavra não aceito pagar r$2100 por este livro pois estava informando q era gratuitoprezados acredito que houve um erro pois descontaram o valor deste livro no meu cartão de crédito quando escolhi este livro foi no dia 30 de março no momento de uma campanha da empresa de doação de ebooks e este livro era uns dos livros disponíveis gratuitamente espero uma solução rápida pois não gostaria de ser lesadoVeio amassado nas pontasoh amazon por favor atualiza essa edição está sem condição de ler faltam frases dos personage

In [86]:
#estudo de frequência 
serie_geral = pd.Series(livro_geral.split())
tabela_geral = serie_geral.value_counts()
tabela_geral

de               828
o                789
que              778
e                747
a                712
                ... 
curtia             1
encosta            1
contraditório      1
estudioso          1
psicológica        1
Length: 5364, dtype: int64

In [87]:
serie_conteudo = pd.Series(livro_conteudo.split())
tabela_conteudo = serie_conteudo.value_counts()
tabela_conteudo

de             621
que            555
e              525
o              515
a              472
              ... 
21               1
repassei         1
suposta          1
devolvi          1
psicológica      1
Length: 4222, dtype: int64

In [88]:
serie_entrega = pd.Series(livro_entrega.split())
tabela_entrega = serie_entrega.value_counts()
tabela_entrega

o                 274
a                 240
que               223
e                 222
de                207
                 ... 
painel              1
conspiração         1
mencionadas         1
personalidades      1
persistente         1
Length: 2042, dtype: int64

In [89]:
#suavização de laplace:


D = len(livro_geral.split()) + len(universo)
def suavizador_laplace(palavra, tabela):
    frequencia_suavizada = dict()
    if palavra in tabela:
        N = tabela[palavra]
        if palavra not in frequencia_suavizada:
            frequencia_suavizada[palavra] = N/D

        else:
            frequencia_suavizada[palavra] += N/D
    else:
        N = 1
        if palavra not in frequencia_suavizada:
            frequencia_suavizada[palavra] = N/D

        else:
            frequencia_suavizada[palavra] += N/D
    return frequencia_suavizada




    

In [63]:
df_test


Mensagem    Target
0    a narrativa é muito lenta e você fica ansioso ...  conteúdo
1    simadorei o livro vou continuar a comprar o re...  conteúdo
2    desculpem mas achei deprimente e não me ensino...  conteúdo
3    preciso saber se essa compra é do livro físico...   entrega
4    embora tenha gostado do livro não faz sentido ...  conteúdo
..                                                 ...       ...
245  livro ruim comparado com seu antecessor a auto...  conteúdo
246  coaching virou um grande modismo assim como cr...  conteúdo
247  nao gostei nem da formatacao nem do conteudo u...  conteúdo
248  para quem tem um mínimo de conhecimento em fin...  conteúdo
249  primeiramente é super previsível e o romance n...  conteúdo

[250 rows x 2 columns]

In [64]:
prob_entrega = len(serie_entrega)/len(serie_geral)
prob_conteudo = len(serie_conteudo)/len(serie_geral)

In [65]:
# mensagens = df_test['Mensagem']
# for i,msg in enumerate(mensagens):
#     for palavra in msg.split():
#         suavizador_laplace(palavra, df_test)
#testando


In [83]:
def classificador(frase, tabela):
    p_frase_tal_tabela  = 1
    for palavra in frase.split():
        p_frase_tal_tabela *= 


In [84]:
mensagens = df_test['Mensagem']
Target = df_test['Target']
dic_de_frases = dict()
for msg in (mensagens):
    for palavra in msg.split():
        suavizador_laplace(palavra, df_test)
for msg in (mensagens):
    dic_de_frases[msg] = classificador(msg, df_test)


In [85]:
dic_de_frases

{'a narrativa é muito lenta e você fica ansioso pra chegar no final e quando chega nada demais acontece não gostei': 1.388027981251613e-94,
 'simadorei o livro vou continuar a comprar o resto da coleção muito emocionante esta sagamelhor di que a minisserie da tv': 1.388027981251613e-94,
 'desculpem mas achei deprimente e não me ensinou muita coisa não gosto de finais clichês mas acho que faltou dar continuidade à história do jeito que acabou foi apenastriste': 2.438492100399387e-130,
 'preciso saber se essa compra é do livro físico pq não quero digital acho que não entendi a propaganda desculpa msm fui no impulso se não for físico quero cancelar a compra': 9.525604098168177e-144,
 'embora tenha gostado do livro não faz sentido também do felipe neto que nos faz ter uma visão de alguém que batalha dentro do youtube o livro felipe neto a trajetória de um dos maiores booktubers é fraquíssimo não acrescenta nada deseduca crianças e adolescentes com brincadeiras do tipo trepa casa ou passa n

___
### Concluindo

___
### Qualidade do Classificador a partir de novas separações dos tweets entre Treinamento e Teste

Caso for fazer esse item do Projeto

___
## Aperfeiçoamento:

Trabalhos que conseguirem pelo menos conceito B vão evoluir em conceito dependendo da quantidade de itens avançados:

* IMPLEMENTOU outras limpezas e transformações que não afetem a qualidade da informação contida nos tweets. Ex: stemming, lemmatization, stopwords
* CONSIDEROU mais de duas categorias na variável Target e INCREMENTOU a quantidade de notícias, mantendo pelo menos 250 notícias por categoria (OBRIGATÓRIO PARA TRIOS, sem contar como item avançado)
* Para Target com duas categorias: CRIOU pelo menos quatro categorias intermediárias de relevância baseadas na probabilidade: ex.: muito relevante, relevante, neutro, irrelevante, muito irrelevante
* EXPLICOU porquê não pode usar o próprio classificador para gerar mais amostras de treinamento
* PROPÔS diferentes cenários para Naïve Bayes fora do contexto do projeto (pelo menos dois cenários, exceto aqueles já apresentados em sala pelos professores: por exemplo, filtro de spam)
* SUGERIU e EXPLICOU melhorias reais com indicações concretas de como implementar (indicar como fazer e indicar material de pesquisa)
* FEZ o item Qualidade do Classificador a partir de novas separações das Notícias entre Treinamento e Teste descrito no enunciado do projeto (OBRIGATÓRIO para conceitos A ou A+)

___
## Referências

[Naive Bayes and Text Classification](https://arxiv.org/pdf/1410.5329.pdf)  **Mais completo**

[A practical explanation of a Naive Bayes Classifier](https://monkeylearn.com/blog/practical-explanation-naive-bayes-classifier/) **Mais simples**